mayastor/src/bdev/nexus/nexus_io.rs

```diff
@@ -1,471 +1,478 @@
+                        inflight += 1;
+                    })
+                })
+            }
+            IoType::WriteZeros => {
+                self.inner_channel().writers.iter().try_for_each(|h| {
+                    self.submit_write_zeroes(h).map(|_| {
+                        inflight += 1;
+                    })
+                })
+            }
+            // we should never reach here, if we do it is a bug.
+            _ => unreachable!(),
+        }
+        .map_err(|se| {
+            match se {
+                Errno::ENOMEM => status = IoStatus::NoMemory,
+                _ => status = IoStatus::Failed,
+            }
+            debug!(
+                "IO submission failed {} already submitted {}",
+                se, inflight
             );
+            se
+        });
 
-            slice[0].iov_base.is_null()
+        if inflight != 0 {
+            self.ctx_as_mut().in_flight = inflight;
+            self.ctx_as_mut().status = status;
+        } else {
+            // if no IO was submitted at all, we can fail the IO now.
+            if matches!(result, Err(Errno::ENOMEM)) {
+                self.no_mem();
+                // right now this can only be EINVAL
+            } else {
+                self.fail();
+            }
         }
-    }
 
-    pub(crate) fn as_ptr(&self) -> *mut spdk_bdev_io {
-        self.0.as_ptr()
+        result
     }
-}
 
-impl Debug for Bio {
-    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
-        writeln!(
-            f,
-            "bdev: {} offset: {:?}, num_blocks: {:?}, type: {:?} status: {:?}, {:p} ",
-            self.bdev_as_ref().name(),
-            self.offset(),
-            self.num_blocks(),
-            self.io_type(),
-            self.status(),
-            self
-        )
+    /// Retire a child for this nexus.
+    async fn child_retire(nexus: String, child: Bdev) {
+        match nexus_lookup(&nexus) {
+            Some(nexus) => {
+                if let Some(child) = nexus.child_lookup(&child.name()) {
+                    let current_state = child.state.compare_and_swap(
+                        ChildState::Open,
+                        ChildState::Faulted(Reason::IoError),
+                    );
+
+                    if current_state == ChildState::Open {
+                        warn!(
+                            "core {} thread {:?}, faulting child {}",
+                            Cores::current(),
+                            Mthread::current(),
+                            child,
+                        );
+
+                        nexus.pause().await.unwrap();
+                        nexus.reconfigure(DrEvent::ChildFault).await;
+                        // TODO: an error can occur here if a separate task,
+                        // e.g. grpc request is also deleting the child.
+                        if let Err(err) =
+                            bdev_destroy(&child.name.clone()).await
```

Comment:
Would need reapply the fix for #794, depending on which is merged first.

Reply:
Yes, its a minor one so we will see, not to worried about that :)